File: ArtificialIntelligence/Astar/pacman_dfs.cpp
// https://www.hackerrank.com/challenges/pacman-dfs
// g++ pacman_dfs.cpp -std=c++1y -g

#include <iostream>
#include <vector>
#include <stack>
#include <set>

#define FORBIDDEN '%'
#define VALID '-'

typedef struct 
{
    int r;
    int c;
}coordinates;

// UP:    [i-1] [j]
// LEFT   [i] [j-1]
// RIGHT  [i] [j+1]
// DOWN   [i+1] [j]
// Print the number of nodes explored ( a node is marked explored only when the node is popped out of the stack ) 


void dfs( int r, int c, int pacman_r, int pacman_c, int food_r, int food_c, std::vector <std::string> grid)
{
	std::vector<coordinates> to_print;
	std::set<coordinates> visited; std::set<coordinates>::iterator it;
	std::stack<coordinates> mystack;

    coordinates init;
    init.r = pacman_r;
    init.c = pacman_c;

    mystack.push(init);

    while(! mystack.empty())
    {
        coordinates v = mystack.top();
        to_print.push_back(v);
        if (v.r = food_r && v.c == food_c)
            break;
        mystack.pop();
        

        it = visited.find(v);
        

    }
    
}


int main(void)
{
    int r,c, pacman_r, pacman_c, food_r, food_c;

    std::cin >> pacman_r >> pacman_c;
    std::cin >> food_r >> food_c;
    std::cin >> r >> c;

    std::vector<std::string> grid;

    for(int i=0; i<r; i++)
    {
    	std::string s; std::cin >> s;
    	grid.push_back(s);
    }

    dfs( r, c, pacman_r, pacman_c, food_r, food_c, grid);

    return 0;
}

File: ArtificialIntelligence/Astar/pacman_dfs.py
# https://www.hackerrank.com/challenges/pacman-dfs

def dfs( r, c, pacman_r, pacman_c, food_r, food_c, grid):
    stack = []
    stack.append((pacman_r, pacman_c))
    explored = []
    path = {}
    path[pacman_r, pacman_c] = (-1, -1)
    while len(stack) > 0:
        v = stack.pop()

        if v not in explored:
            explored.append(v)
            
            if v[0] - 1 >= 0 and grid[v[0]-1][v[1]] != '%': # UP
                stack.append((v[0]-1, v[1]))
                if (v[0]-1, v[1]) not in path: path[(v[0]-1, v[1])] = v
                
            if v[1] - 1 >= 0 and grid[v[0]][v[1] - 1] != '%': # LEFT
                stack.append((v[0], v[1]-1))
                if (v[0], v[1]-1) not in path: path[(v[0], v[1]-1)] = v
                
            if v[1] + 1 < c and grid[v[0]][v[1] + 1] != '%': # RIGHT
                stack.append((v[0], v[1]+1))
                if (v[0], v[1]+1) not in path: path[(v[0], v[1]+1)] = v
                
            if v[0] + 1 < r and grid[v[0] + 1][v[1]] != '%': # DOWN
                stack.append((v[0]+1, v[1]))
                if (v[0]+1, v[1]) not in path: path[(v[0]+1, v[1])] = v

            if v[0] == food_r and v[1] == food_c:
                break

    print(len(explored) )
    for i in explored:
        print(i[0], i[1])

    order = []
    father = (food_r, food_c)
    while father != (-1, -1):
        order.append(father)
        father = path[father]

    i = len(order) -1
    print(i)
    while i >= 0:
        print(order[i][0], order[i][1])
        i-=1




if __name__ == "__main__":
    pacman_r, pacman_c = [ int(i) for i in input().strip().split() ]
    food_r, food_c = [ int(i) for i in input().strip().split() ]
    r,c = [ int(i) for i in input().strip().split() ]

    grid = []
    for i in range(0, r):
        grid.append(input().strip())

    dfs(r, c, pacman_r, pacman_c, food_r, food_c, grid)